Core: Declarative story definition in main.js

ndelangen opened the pull request:
Issue: preview.js is required to inject stories, this is in imperative code, and impossible to extract. If we declaratively configure the stories in main.js, we could determine the stories without running storybook.

## What I did
Make it possible to define stories entries in main.js

Hopefully with enough minor steps like this we can make it less and less likely users need preview.js ever.

ndelangen committed:
Make it possible to define stories entries in main.js

lowering the need for preview.js

ndelangen committed:
ADD support for globs

ndelangen committed:
CHANGE make it possible for presets values to be arrays & objects & primitives

Arrays & objects are merged, primitives replace old values

ndelangen committed:
CHANGe main.js to make use of presets that can be primitives

ndelangen committed:
FIX incorrect entries injection

tmeasday commented:
@ndelangen this is awesome!

I'm not sure if it should be in this PR or seperately, but we need to update storyshots to support this. It'll be great though. @shilman and I discussed it a bit, and here's what should happen:

Right now to use SS:

1. You write a `.test.js` file that imports and calls `initStoryshots`
2. `initStoryshots` ends up `require`-ing your `.storybook/config.js`
3. You `.storybook/config.js`, which itself:
  a) does whatever configuration (adding decorators etc)
  b) then calls [`requireContext()`](https://www.npmjs.com/package/require-context.macro) a `require.context()` macro that in Jest actually just ends up running a bunch of `require()` calls.

This will actually simplify things and remove the need for that macro which would be awesome! To support this properly, we'd need to:

1. Support triconfig properly in storyshots. I think this probably just means also requiring `.storybook/preview.js` in step 2. above.

2. After doing step 2-3 above, `initStoryshots` should then: require `.storybook/main.js`, pull the `stories` key out, then require all the stories.

ndelangen commented:
@tmeasday I propose to tackle that in another PR.

If we release this without storyshots support, users can still use the 'old way of doing things'

ndelangen committed:
Merge branch 'next' into core/magic-entry-storyshots

ndelangen committed:
CHANGE html to use new triconfig

ndelangen committed:
ADD compatibility for triconfig to storyshots

ndelangen commented:
@tmeasday https://github.com/storybookjs/storybook/pull/8765

ndelangen committed:
Merge branch 'next' into core/magic-entry

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

shilman reviewed (approved):
This looks fantastic @ndelangen 👏👏👏 

Would love to go through it together when you're available and also get feedback from others, since it's a major change

shilman reviewed (approved):
This looks fantastic @ndelangen 👏👏👏 

Would love to go through it together when you're available and also get feedback from others, since it's a major change

ndelangen committed:
FIX review comments, CHANGE to only support a single preview.*

ndelangen committed:
Merge branch 'next' into core/magic-entry

shilman commented on lib/core/src/server/preview/iframe-webpack.config.js:
```diff
@@ -44,6 +71,16 @@ export default ({
+      stories && stories.length
+        ? new VirtualModulePlugin({
+            [path.resolve(path.join(configDir, `generated-entry.js`))]: `
+              import { configure, addDecorator, addParameters } from '@storybook/react';
```
Does this need to change to `@storybook/${framework}`??

shilman replied:
Framework is currently used in all the presets from `options.framework` 😉

ndelangen replied:
yes

ndelangen replied:
✅

shilman reviewed (changes requested):
Seems to be react-specific at this point

shilman commented:
Also can you upgrade one more non-react example to test?

ndelangen committed:
Merge branch 'next' into core/magic-entry

# Conflicts:
#	examples/official-storybook/main.js

ndelangen committed:
Merge branch 'core/magic-entry' of github.com:storybooks/storybook into core/magic-entry

ndelangen committed:
FIX critical issue where framework was assumed to be 'react'

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

# Conflicts:
#	addons/storyshots/storyshots-core/package.json
#	examples/html-kitchen-sink/.storybook/presets.js

ndelangen committed:
FIX bug with resolving of preview.js (doh)

ndelangen commented:
> Also can you upgrade one more non-react example to test?

I can't without adding storyshots support

I have migrated the html example in this PR https://github.com/storybookjs/storybook/pull/8765

ndelangen committed:
Merge branch 'next' into core/magic-entry

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

# Conflicts:
#	addons/storyshots/storyshots-core/package.json

ndelangen committed:
CHANGE to use configure over copied loadStories fn

ndelangen committed:
ADD support for configure to accept an array of string pointing to files, without require.context

ndelangen committed:
CLEANUP && Use configure function from clientApi

ndelangen committed:
CLEANUP

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

ndelangen committed:
ROLLBACK the change to allow for any string to be loadable, as this is nightmare-scenario for webpack
"Critical dependency: the request of a dependency is an expression"

Instead allow the array to contain "functions that do the require & return exports" or "the exports directly"

ndelangen commented:
@shilman is this all good to go?

ndelangen committed:
tom, you're correct as always!

ndelangen committed:
Merge branch 'next' into core/magic-entry

# Conflicts:
#	yarn.lock

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

shilman committed:
Merge pull request #8765 from storybookjs/core/magic-entry-storyshots

Storyshots: Triconfig support

ndelangen commented:
EPIC WIN
